include/odp/api/spec/packet_io_types.h

```diff
@@ -1086,6 +1092,18 @@ typedef struct odp_pktio_capability_t {
 
 } odp_pktio_capability_t;
 
+/** Parameters for ODP_LSO_WRITE_BITS custom operation */
+typedef struct odp_lso_write_bits_t {
+	/** Bitmask to select which bits to write */
+	uint8_t mask;
```

Comment:
Change to uint16_t to support 16-bit flags.

Reply:
uint16_t seems quite arbitrary. Why not then uint32_t or uint64_t? I think uint8_t keeps the API simpler and still allows 16-bit flags through the use of two WRITE_BITS operations.

Reply:
The IPv4 16-bit fragment offset is shared with flags. Hardware (HW) may modify both fields in a single operation, so the flags will be 16-bit. This will save the number of operations in a custom profile, which are limited in HW. Even though it is 16-bit, based on size, it may modify a single byte as well.

Reply:
The way things are done in the HW is different from the way things are expressed in the API. The ODP implementation can easily combine multiple operations into one and change their width if that is more efficient in the HW.

Reply:
So let's make this as follows(some thing similar), so that it is flexible and easy to use for different implementations 
uint8_t mask[8];
uint8_t value[8];
and based on size argument we will select number of bytes from this array. As per spec max valid value for size was 8, so make this array to 8 bytes.

Reply:
We can consider that. But it is not easier for implementations than the fixed 1 byte width, since now the implementation does not only have to handle 1 byte widths and map them to HW but it also has to handle all the other widths. So the rationale for the change cannot be ease of implementation but rather convenience of the user of the API or cleanliness of the API (since the sizeis a common parameter and not an operation specific parameter, which it could also be).

Reply:
It is easy to implement, and mapping multi-byte to hardware is simpler compared to the previous method. Regarding larger widths, if the hardware supports only X bytes, then sizes greater than X will return -ENOTSUP.

Reply:
> mapping multi-byte to hardware is simpler compared to the previous method

It is not simpler but more difficult. The proposed method is essentially the same as your proposal with the additional restriction that size must always be 1 byte in the WRITE_BITS operation. Giving more widths for the user the select from does not make the mapping easier but potentially more difficult. Even if we supported multiple widths, the API user could still use 1-byte operation (e.g. in IPv4 the fragmentation related flags are contained in one byte) and you would need to map that to a 16-bit operation if that is all your HW has.

IOW, ODP API is an abstraction to enable the user to tell what kind of LSO behaviour she wants. ODP implementation then maps that to the HW if at all possible. It is no ok to unnecessarily expose HW details in the API (e.g. have byte-wide WRITE_BITS operation fail in some ODP implementation and version just because the HW has only 16-bit wide operations natively).

> Regarding larger widths, if the hardware supports only X bytes, then sizes greater than X will return -ENOTSUP.

No, the ODP implementation is supposed to map the LSO profile to the HW if it is possible to do so. So larger widths in the profile would correspond to multiple narrower operations in the HW. Profile creation should fail only if it is not possible to do the mapping (e.g. there is only one 8-bit operation in the HW and the LSO requests a wider operation with mask bits set in more than 1 byte).

Reply:
> How would the API know that the first 2 bytes are one operation and the remaining 4 bytes are another operation?

There is no 2-byte and 4-byte operation there, but just a bunch of bits that need to be changed. The way the bit changes are split to some HW primitives is an implementation detail that does not concern the API. From the API point-of-view there is only one operation, the segmentation operation. The API just tells unambiguously which bits to change and how, the ODP implementation then maps the request to whatever update primitives it needs to use with the underlying HW. Only the externally observable behaviour matters, so it does not make any difference whether the implementation maps the request to one 2-byte HW op and one 4-byte op or vice versa or something else.

> As I proposed, if you change this to an array of masks and values, it becomes very easy to map for all 1-byte, 2-byte, 4-byte, and 8-byte combinations.

How does it become easier? In the above example, the user could still configure the profile with 6 1-byte operations and the implementation would have to make exactly the same mappings to the 2 and 4 byte HW ops. We do not want to have profile creation fail if the width used there does not match whatever happens to be the width of the HW primitive as that would unnecessarily expose implementation details to the API user and would make using the API more difficult.

IOW, if we change the API to allow more ways to configure the same thing, the implementation does not become easier but more difficult. So ease of implementation cannot be reason to make the API larger. The implementation become easier only if you add restriction in the API, but we do not want to add HW-dependent restrictions which can be avoided by a bit of logic in the implementation.

Reply:
Just visualize it from a header modification point of view. On platform-B, we need to perform two modifications:
1. At X-offset, we need to modify a 16-bit header field.
2. At Y-offset, we need to modify a 32-bit header field.

In my proposal, it is simple with two operations:
1. Operation: write_bits
   - Offset: X 
   - Size: 2 
   - Mask: [2]
   - Value: [2]

2. Operation: write_bits
   - Offset: Y
   - Size: 4
   - Mask: [4]
   - Value: [4]

How does your proposal support this? This is a valid test case for header modifications.

Reply:
The basis for my proposal is that combining bytes into a single operation is driven by the header fields of protocols rather than hardware requirements.

Reply:
> How does your proposal support this?

Through 2 or more byte-wide write_bits operations. What is the problem?

Reply:
Assume (X = 10) and (Y = 12) in the above example. Could you describe how your 6 operations looks like?

Reply:
> Assume (X = 10) and (Y = 12) in the above example. Could you describe how your 6 operations looks like?

That would be configured through an LSO profile with ODP_LSO_PROTO_CUSTOM lso_proto and with as many write_bits operations as there are bytes that need to be changed. If you need to change bits in all of the 6 bytes in your "header fields", then you would need to have 6 custom fields in the LSO profile, each field containing a write_bits operation for one of the bytes (i.e. the offsets in the ops would be 10, 11, 12, …, 15).

> because from protocol POV it is 2 field changes not 48 bit (6-byte) operations

We do not care about the protocol point-of-view. Custom LSO profile is not aware and does not need to be aware of the protocol carried in the packet or whatever 'fields' the protocol might define. For LSO purposes (with custom protocol) the packet is just a sequence of bytes that we modify using the modify operations the way we wish. The only thing LSO needs to know is which part of the packet is considered header and which payload for segmentation purposes, but that is not related to the write_bits operation.

> In my proposal, it is simple with two operations:

Yes, in this particular case it might be simpler for the API user if there are to-be-modified bits in more than one byte in each of your "header fields". But for an ODP implementation it would be a larger API and therefore more difficult to implement, not less, as you claimed before.

Reply:
As discussed offline, the rationale for having multiple widths for the write_bits operation cannot be ease of implementation (of ODP) since with multiple widths the implementation would still have to be able 1-byte wide write_bits if the application happens to use them. Multiple widths, while not functionally adding anything, might be more convenient for some applications.

Another issue is the maximum number of custom fields, which one might face sooner with 1-byte wide operations than with wider operations. But maybe that is not a real concern in practice at this point. If we do not support multiple widths, maybe we specify that the size parameter has to be one, to make it easier to add wider write_bits operation later in a backward compatible manner.